code/game/gamemodes/malfunction/Malf_Modules.dm

```diff
@@ -1,51 +1,305 @@
+	if(owner_AI && owner_AI.malf_cooldown > world.time)
+		return
+
+/datum/action/innate/ai/Trigger()
+	. = ..()
+	if(auto_use_uses)
+		adjust_uses(-1)
+	if(cooldown_period)
+		owner_AI.malf_cooldown = world.time + cooldown_period
+
+/datum/action/innate/ai/proc/adjust_uses(amt, silent)
+	uses += amt
+	if(!silent && uses)
+		to_chat(owner, "<span class='notice'>[name] now has <b>[uses]</b> use[uses > 1 ? "s" : ""] remaining.</span>")
+	if(!uses)
+		if(initial(uses) > 1) //no need to tell 'em if it was one-use anyway!
+			to_chat(owner, "<span class='warning'>[name] has run out of uses!</span>")
+		Remove(owner)
+		qdel(src)
+
+//Framework for ranged abilities that can have different effects by left-clicking stuff.
+/datum/action/innate/ai/ranged
+	name = "Ranged AI Action"
+	auto_use_uses = FALSE //This is so we can do the thing and disable/enable freely without having to constantly add uses
+	var/obj/effect/proc_holder/ranged_ai/linked_ability //The linked proc holder that contains the actual ability code
+	var/linked_ability_type //The path of our linked ability
+
+/datum/action/innate/ai/ranged/New()
+	if(!linked_ability_type)
+		WARNING("Ranged AI action [name] attempted to spawn without a linked ability!")
+		qdel(src) //uh oh!
+		return
+	linked_ability = new linked_ability_type()
+	linked_ability.attached_action = src
+	..()
+
+/datum/action/innate/ai/ranged/adjust_uses(amt, silent)
+	uses += amt
+	if(!silent && uses)
+		to_chat(owner, "<span class='notice'>[name] now has <b>[uses]</b> use[uses > 1 ? "s" : ""] remaining.</span>")
+	if(!uses)
+		if(initial(uses) > 1) //no need to tell 'em if it was one-use anyway!
+			to_chat(owner, "<span class='warning'>[name] has run out of uses!</span>")
+		Remove(owner)
+		QDEL_IN(src, 100) //let any active timers on us finish up
+
+/datum/action/innate/ai/ranged/Destroy()
+	QDEL_NULL(linked_ability)
+	return ..()
+
+/datum/action/innate/ai/ranged/Activate()
+	linked_ability.toggle(owner)
+	return TRUE
+
+//The actual ranged proc holder.
+/obj/effect/proc_holder/ranged_ai
+	var/enable_text = "<span class='notice'>Hello World!</span>" //Appears when the user activates the ability
+	var/disable_text = "<span class='danger'>Goodbye Cruel World!</span>" //Context clues!
+	var/datum/action/innate/ai/ranged/attached_action
+
+/obj/effect/proc_holder/ranged_ai/Destroy()
+	QDEL_NULL(attached_action)
+	return ..()
+
+/obj/effect/proc_holder/ranged_ai/proc/toggle(mob/user)
+	if(active)
+		remove_ranged_ability(disable_text)
+	else
+		add_ranged_ability(user, enable_text)
+
+
+//The datum and interface for the malf unlock menu, which lets them choose actions to unlock.
+/datum/module_picker
+	var/temp
+	var/processing_time = 50
+	var/list/possible_modules
+
+/datum/module_picker/New()
+	possible_modules = list()
+	for(var/type in typesof(/datum/AI_Module))
+		var/datum/AI_Module/AM = new type
+		if((AM.power_type && AM.power_type != /datum/action/innate/ai) || AM.upgrade)
+			LAZYADD(possible_modules, AM)
```

Comment:
I always thought the lazy macros were just generally preferred. Are there certain cases in which they're used, and certain cases in which they're not?